// xstream/extra/split.js
"use strict";
var core_1 = require('../core');
var SeparatorIL = (function () {
    function SeparatorIL(out, op) {
        this.out = out;
        this.op = op;
    }
    SeparatorIL.prototype._n = function (t) {
        this.op.up();
    };
    SeparatorIL.prototype._e = function (err) {
        this.out._e(err);
    };
    SeparatorIL.prototype._c = function () {
        this.op.curr._c();
        this.out._c();
    };
    return SeparatorIL;
}());
var SplitOperator = (function () {
    function SplitOperator(s, // s = separator
        ins) {
        this.s = s;
        this.ins = ins;
        this.type = 'split';
        this.curr = new core_1.Stream();
        this.out = null;
        this.sil = core_1.emptyListener; // sil = separator InternalListener
    }
    SplitOperator.prototype._start = function (out) {
        this.out = out;
        this.s._add(this.sil = new SeparatorIL(out, this));
        this.ins._add(this);
        out._n(this.curr);
    };
    SplitOperator.prototype._stop = function () {
        this.ins._remove(this);
        this.s._remove(this.sil);
        this.curr = new core_1.Stream();
        this.out = null;
        this.sil = core_1.emptyListener;
    };
    SplitOperator.prototype.up = function () {
        this.curr._c();
        this.out._n(this.curr = new core_1.Stream());
    };
    SplitOperator.prototype._n = function (t) {
        if (!this.out)
            return;
        this.curr._n(t);
    };
    SplitOperator.prototype._e = function (err) {
        var u = this.out;
        if (!u)
            return;
        u._e(err);
    };
    SplitOperator.prototype._c = function () {
        var u = this.out;
        if (!u)
            return;
        this.curr._c();
        u._c();
    };
    return SplitOperator;
}());
exports.SplitOperator = SplitOperator;
function split(separator) {
    return function splitOperator(ins) {
        return new core_1.Stream(new SplitOperator(separator, ins));
    };
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = split;
