// @cycle/dom/lib/render-dom.js
"use strict";

var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; })();

var Rx = require("rx");
var fromEvent = require("./fromevent");
var VDOM = {
  h: require("./virtual-hyperscript"),
  diff: require("virtual-dom/diff"),
  patch: require("virtual-dom/patch"),
  parse: typeof window !== "undefined" ? require("vdom-parser") : function () {}
};

var _require = require("./transposition");

var transposeVTree = _require.transposeVTree;

var matchesSelector = undefined;
// Try-catch to prevent unnecessary import of DOM-specifics in Node.js env:
try {
  matchesSelector = require("matches-selector");
} catch (err) {
  matchesSelector = function () {};
}

function isElement(obj) {
  return typeof HTMLElement === "object" ? obj instanceof HTMLElement || obj instanceof DocumentFragment : //DOM2
  obj && typeof obj === "object" && obj !== null && (obj.nodeType === 1 || obj.nodeType === 11) && typeof obj.nodeName === "string";
}

function wrapTopLevelVTree(vtree, rootElem) {
  var _vtree$properties$id = vtree.properties.id;
  var vtreeId = _vtree$properties$id === undefined ? "" : _vtree$properties$id;
  var _vtree$properties$className = vtree.properties.className;
  var vtreeClass = _vtree$properties$className === undefined ? "" : _vtree$properties$className;

  var sameId = vtreeId === rootElem.id;
  var sameClass = vtreeClass === rootElem.className;
  var sameTagName = vtree.tagName.toUpperCase() === rootElem.tagName;
  if (sameId && sameClass && sameTagName) {
    return vtree;
  }
  var attrs = {};
  if (rootElem.id) {
    attrs.id = rootElem.id;
  }
  if (rootElem.className) {
    attrs.className = rootElem.className;
  }
  return VDOM.h(rootElem.tagName, attrs, [vtree]);
}

function makeDiffAndPatchToElement$(rootElem) {
  return function diffAndPatchToElement$(_ref) {
    var _ref2 = _slicedToArray(_ref, 2);

    var oldVTree = _ref2[0];
    var newVTree = _ref2[1];

    if (typeof newVTree === "undefined") {
      return Rx.Observable.empty();
    }

    var prevVTree = wrapTopLevelVTree(oldVTree, rootElem);
    var nextVTree = wrapTopLevelVTree(newVTree, rootElem);
    /* eslint-disable */
    rootElem = VDOM.patch(rootElem, VDOM.diff(prevVTree, nextVTree));
    /* eslint-enable */
    return Rx.Observable.just(rootElem);
  };
}

function renderRawRootElem$(vtree$, domContainer) {
  var diffAndPatchToElement$ = makeDiffAndPatchToElement$(domContainer);
  return vtree$.flatMapLatest(transposeVTree).startWith(VDOM.parse(domContainer)).pairwise().flatMap(diffAndPatchToElement$);
}

function isolateSource(source, scope) {
  return source.select(".cycle-scope-" + scope);
}

function isolateSink(sink, scope) {
  return sink.map(function (vtree) {
    var c = (vtree.properties.className + " cycle-scope-" + scope).trim();
    vtree.properties.className = c;
    return vtree;
  });
}

function makeIsStrictlyInRootScope(rootList, namespace) {
  var classIsForeign = function classIsForeign(c) {
    var matched = c.match(/cycle-scope-(\S+)/);
    return matched && namespace.indexOf("." + c) === -1;
  };
  return function isStrictlyInRootScope(leaf) {
    for (var el = leaf.parentElement; el !== null; el = el.parentElement) {
      if (rootList.indexOf(el) >= 0) {
        return true;
      }
      var classList = el.className.split(" ");
      if (classList.some(classIsForeign)) {
        return false;
      }
    }
    return true;
  };
}

function makeEventsSelector(element$) {
  return function events(eventName) {
    var useCapture = arguments.length <= 1 || arguments[1] === undefined ? false : arguments[1];

    if (typeof eventName !== "string") {
      throw new Error("DOM driver's events() expects argument to be a " + "string representing the event type to listen for.");
    }

    return element$.flatMapLatest(function (elements) {
      if (elements.length === 0) {
        return Rx.Observable.empty();
      }
      return fromEvent(elements, eventName, useCapture);
    }).share();
  };
}

function makeElementSelector(rootEl$) {
  return function select(selector) {
    if (typeof selector !== "string") {
      throw new Error("DOM driver's select() expects the argument to be a " + "string as a CSS selector");
    }

    var namespace = this.namespace;
    var scopedSelector = (namespace.join(" ") + " " + selector).trim();
    var element$ = selector.trim() === ":root" ? rootEl$ : rootEl$.map(function (x) {
      var array = Array.isArray(x) ? x : [x];
      return array.map(function (element) {
        if (matchesSelector(element, scopedSelector)) {
          return [element];
        } else {
          var nodeList = element.querySelectorAll(scopedSelector);
          return Array.prototype.slice.call(nodeList);
        }
      }).reduce(function (prev, curr) {
        return prev.concat(curr);
      }, []).filter(makeIsStrictlyInRootScope(array, namespace));
    });
    return {
      observable: element$,
      namespace: namespace.concat(selector),
      select: makeElementSelector(element$),
      events: makeEventsSelector(element$),
      isolateSource: isolateSource,
      isolateSink: isolateSink
    };
  };
}

function validateDOMSink(vtree$) {
  if (!vtree$ || typeof vtree$.subscribe !== "function") {
    throw new Error("The DOM driver function expects as input an " + "Observable of virtual DOM elements");
  }
}

function makeDOMDriver(container) {
  // Find and prepare the container
  var domContainer = typeof container === "string" ? document.querySelector(container) : container;
  // Check pre-conditions
  if (typeof container === "string" && domContainer === null) {
    throw new Error("Cannot render into unknown element `" + container + "`");
  } else if (!isElement(domContainer)) {
    throw new Error("Given container is not a DOM element neither a selector " + "string.");
  }

  return function domDriver(vtree$) {
    validateDOMSink(vtree$);
    var rootElem$ = renderRawRootElem$(vtree$, domContainer).startWith(domContainer).replay(null, 1);
    var disposable = rootElem$.connect();
    return {
      namespace: [],
      select: makeElementSelector(rootElem$),
      dispose: disposable.dispose.bind(disposable),
      isolateSource: isolateSource,
      isolateSink: isolateSink
    };
  };
}

module.exports = {
  isElement: isElement,
  wrapTopLevelVTree: wrapTopLevelVTree,
  makeDiffAndPatchToElement$: makeDiffAndPatchToElement$,
  renderRawRootElem$: renderRawRootElem$,
  validateDOMSink: validateDOMSink,

  makeDOMDriver: makeDOMDriver
};
